File: src/runner/reclaim.zig
//! Lease reclaim — re-leasing an expired holder's event from Postgres alone.
//!
//! When `affinity.claim` wins a zombie whose prior claim had expired, the dead
//! holder's still-`active` lease row carries the durable event envelope + the
//! billing context. `findPriorActive` returns it and `markExpired` retires it,
//! so the caller can re-lease the SAME event under the fresh higher fencing
//! token — no Redis re-read (the envelope is durable in Postgres) and no
//! re-billing (the original lease already debited). If there is no prior active
//! lease the zombie is simply free and the caller takes a fresh event instead.
//!
//! Arena allocator (`hx.alloc`): every returned slice is arena-dup'd and freed
//! when the request ends — see service.zig's module note.

const std = @import("std");
const pg = @import("pg");
const PgQuery = @import("../db/pg_query.zig").PgQuery;
const protocol = @import("protocol.zig");

/// The dead holder's lease: the event envelope to re-lease + the billing
/// context to reuse (no re-charge). All slices arena-dup'd.
pub const PriorLease = struct {
    lease_id: []const u8,
    event_id: []const u8,
    actor: []const u8,
    event_type: []const u8,
    request_json: []const u8,
    event_created_at: i64,
    workspace_id: []const u8,
    tenant_id: []const u8,
    posture: []const u8,
    model: []const u8,
};

/// The zombie's latest `active` lease — the holder whose claim just expired, or
/// null when the zombie is free (no active lease ⇒ the caller takes a fresh
/// event). Called only after `affinity.claim` won, so an active row here is
/// unambiguously the reclaimed holder.
pub fn findPriorActive(conn: *pg.Conn, alloc: std.mem.Allocator, zombie_id: []const u8) !?PriorLease {
    var q = PgQuery.from(try conn.query(
        \\SELECT id::text, event_id, actor, event_type, request_json,
        \\       event_created_at, workspace_id::text, tenant_id::text, posture, model
        \\FROM fleet.runner_leases
        \\WHERE zombie_id = $1::uuid AND status = $2
        \\ORDER BY fencing_token DESC LIMIT 1
    , .{ zombie_id, protocol.RUNNER_LEASE_STATUS_ACTIVE }));
    defer q.deinit();
    const row = try q.next() orelse return null;
    return PriorLease{
        .lease_id = try alloc.dupe(u8, try row.get([]const u8, 0)),
        .event_id = try alloc.dupe(u8, try row.get([]const u8, 1)),
        .actor = try alloc.dupe(u8, try row.get([]const u8, 2)),
        .event_type = try alloc.dupe(u8, try row.get([]const u8, 3)),
        .request_json = try alloc.dupe(u8, try row.get([]const u8, 4)),
        .event_created_at = try row.get(i64, 5),
        .workspace_id = try alloc.dupe(u8, try row.get([]const u8, 6)),
        .tenant_id = try alloc.dupe(u8, try row.get([]const u8, 7)),
        .posture = try alloc.dupe(u8, try row.get([]const u8, 8)),
        .model = try alloc.dupe(u8, try row.get([]const u8, 9)),
    };
}

/// Retire the reclaimed holder's lease so it is no longer the zombie's active
/// lease. Its late report is independently fenced by the bumped token.
pub fn markExpired(conn: *pg.Conn, lease_id: []const u8) !void {
    const now_ms = std.time.milliTimestamp();
    _ = conn.exec(
        \\UPDATE fleet.runner_leases SET status = $2, updated_at = $3 WHERE id = $1::uuid
    , .{ lease_id, protocol.RUNNER_LEASE_STATUS_EXPIRED, now_ms }) catch return error.LeaseExpireFailed;
}

File: rustd/crates/afd_fleet/src/lease/reclaim.rs
//! Re-leasing a dead holder's event from Postgres alone.
//!
//! When a claim wins a fleet whose prior claim had lapsed, the dead holder's
//! still-`active` lease row carries the billing context and the event row it
//! names carries the body. [`Leases::reclaim_prior_active`] takes both in one
//! atomic statement and expires the old lease on the way, so the caller can
//! re-lease the SAME event under the fresh, higher fence — no Redis re-read
//! (the envelope is durable in Postgres) and no re-billing (the original lease
//! already debited).
//!
//! No prior active lease means the fleet is simply free, and the caller pulls a
//! fresh event instead.

use afd_core::clock::UnixMillis;
use afd_core::id::Uuid7;
use sqlx::Row as _;

use crate::error::{Result, query};
use crate::lease::store::Leases;
use crate::lease::sql;

/// Statement name, for the context a query failure carries.
const CONTEXT_RECLAIM: &str = "lease reclaim";

/// The billing a reclaimed lease carries FORWARD rather than re-charging.
///
/// Separate from [`Reclaimed`] because it is the half that must not be
/// re-derived: the original lease already debited the tenant, so a reclaim that
/// resolved these afresh could bill twice, and a rotation between the two
/// resolutions would bill the wrong rate. Carried, never recomputed.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Reused {
    /// The tenant the original lease debited.
    pub tenant_id: String,
    /// The billing posture it was charged under.
    pub posture: String,
    /// The model it was charged for.
    pub model: String,
}

/// A dead holder's work, re-leasable under a fresh fence.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Reclaimed {
    /// The lease that was expired to release this work.
    pub lease_id: String,
    /// The event to re-run.
    pub event_id: String,
    /// Who raised it.
    pub actor: String,
    /// Its type.
    pub event_type: String,
    /// Its body, as stored.
    pub request_json: String,
    /// When the producer raised it.
    pub event_created_at: i64,
    /// The workspace it belongs to.
    pub workspace_id: String,
    /// The billing carried forward.
    pub reused: Reused,
}

impl Leases {
    /// Expire `fleet_id`'s latest active lease and take back what it was
    /// executing.
    ///
    /// Answers `None` when the fleet has no active lease — it is simply free,
    /// and the caller pulls fresh work. Also answers `None` when the event row
    /// has been deleted out from under a live lease: the statement's join is
    /// INNER, so there is nothing to re-deliver, and the dead lease is still
    /// expired on the way past rather than left lingering.
    ///
    /// Call only after winning the claim, so the row found here is
    /// unambiguously the holder this runner displaced.
    ///
    /// # Errors
    /// Reports a datastore that would not answer.
    pub async fn reclaim_prior_active(
        &self,
        fleet_id: &Uuid7,
        now: UnixMillis,
    ) -> Result<Option<Reclaimed>> {
        let mut connection = self.pool().acquire().await?;
        let found = sqlx::query(sql::lease::RECLAIM_PRIOR_ACTIVE)
            .bind(fleet_id.as_str())
            .bind(sql::LEASE_STATUS_ACTIVE)
            .bind(sql::LEASE_STATUS_EXPIRED)
            .bind(now.as_millis())
            .fetch_optional(&mut *connection)
            .await
            .map_err(query(CONTEXT_RECLAIM))?;

        let Some(row) = found else {
            return Ok(None);
        };
        let column = |index| {
            row.try_get::<String, _>(index)
                .map_err(query(CONTEXT_RECLAIM))
        };
        Ok(Some(Reclaimed {
            lease_id: column(0)?,
            event_id: column(1)?,
            actor: column(2)?,
            event_type: column(3)?,
            request_json: column(4)?,
            event_created_at: row.try_get(5).map_err(query(CONTEXT_RECLAIM))?,
            workspace_id: column(6)?,
            reused: Reused {
                tenant_id: column(7)?,
                posture: column(8)?,
                model: column(9)?,
            },
        }))
    }
}
